kernel/src/scan/mod.rs

```diff
@@ -107,11 +121,15 @@ pub struct ScanResult {
     pub mask: Option<Vec<bool>>,
 }
 
-/// Scan uses this to set up what kinds of columns it is scanning. A reference to the field is
-/// stored so we can get the name and data_type out when building the read expression
-pub enum ColumnType<'a> {
-    Selected(&'a StructField),
-    Partition(&'a StructField),
+/// Scan uses this to set up what kinds of columns it is scanning. For `Selected` we just store the
+/// name of the column, as that's all that's needed during the actual query. For `Partition` we
+/// store an index into the logical schema for this query since later we need the data type as well
+/// to materialize the partition column.
+pub enum ColumnType {
+    // A column, selected from the data, as is
+    Selected(String),
```

Comment:
If I'm not mistaken, this could be `&'a str` since it's either a reference to the field's name, or to an entry from the field's metadata map? That way, we could avoid creating a `String` unless partitions and/or CMM are actually involved.

Reply:
I still keep a `Vec<ColumnType>` in `Scan` (the `all_fields` field). This is to avoid having to recompute everything each time.

If we make this a reference with a lifetime, then `Scan` has to have a lifetime, and then everything gets... messy.

I'd actually argue for going the "other way" if we want to remove string allocation, and just have this also store an index in the schema.

Reply:
Ah, I forgot that you hoisted `all_fields` to be a field computed previously, where it used to be converted internally...

However -- I don't think we can "just" store a schema index for the column name, when it could be name-mapped or not?